# Number system

Number system gives us a way of describing a number.
The same number can be described differently in different systems.

The **radix/base** of a number system determines:
* the largest possible symbol in the system
* the value of each digit as we move across the positions.

Digits to the left of the **radix point** have a positive powered base as the value, while those to the right has a negative powered base.

Given a radix $r$, a number in the number system of that radix is defined as:
$$A_r = (a_n a_{n-1} \dots a_0 . a_{-1} \dots a_{-m})_r = \sum _{i=-m}^n a_i r^i \quad a_i \in \{0, \dots r-1\}
$$

## Decimal system

This is the most used system in real life.
This has a radix of 10.
The digit ranges from 0-9.

For example, $(13.75)_{10}$ is a decimal number. 
The subscript is there to indicate to us the number system the number is being represented in.

Using the fundamental of the decimal system, we can see that $13.75 = 1 \times 10^1 + 3 \times 10^0 + 7 \times 10^{-1} + 5 \times 10^{-2}$.

## Binary system

This is the most common system in the digital world.
It has a radix of 2, and thus the digit are simply 0 or 1.

Consider the binary number $(1101.11)_2$.
The value of this number in decimal is:
$$
1 \times 2^3 + 1 \times 2^2 + 1 \times 2 ^ 0 + 1 \times 2 ^{-1} + 1 \times 2^{-2} = (13.75)_{10}
$$

Hence, even though it looks different, it is the same as $13.75$ in decimal.

## Hexadecimal system

It has a radix of 16, and thus the digit ranges from 0-15.
To represent digits from 10-15, we use the letters A-F.

Consider the hexadecimal number $(D.C)$.
This is equals to $13 \times 16 ^0 + 12 \times 16^{-1} = (13.75)_{16}$.

## Octal system

It has a radix of 8, with digits that ranges from 0-7.

Consider the octal number $(15.6)$.
This is equals to $1 \times 8^1 + 5 \times 8^0 + 6 \times 8^{-1} = 13.75$.

Note that some decimal number cannot be represented as a finite digit number in another system.
For example, $(0.1)_{10} = 2^{-1} + 2^{-2} +2^{-3} \dots = (0.\bar 1)_2$, where the bar above the 1 indicates that the number repeats 1 infinitely.

## Conversion

### To decimal

Conversion to decimal is rather straight forward.
Just compute the sum of each digit multiplied by its value to obtain the value in decimal.

In [41]:
def to_dec(string: str, r: int):
    def symbol_val(symbol: chr):
        return '0123456789ABCDEF'.index(symbol)
    
    left, *right = string.split('.')
    if right:
        right = right[0]
        
    total = 0 
    for n, a in enumerate(reversed(left)):
        n, a = int(n), symbol_val(a)
        total += a * r ** n
        
    for n, a in enumerate(right, start=1):
        n, a = int(n), symbol_val(a)
        total += a * r ** -n
    return total
    

In [42]:
print(to_dec('1101.11', 2))
print(to_dec('D.C', 16))
print(to_dec('15.6', 8))

13.75
13.75
13.75


## From decimal

Consider a number $A_r$.
$$A_r = (a_n, a_{n-1} \dots a_0 . a_{-1} \dots a_{-m})_r = \sum _{i=-m}^n a_i r^i 
$$

Now consider the number if we remove the parts after the radix point.
$$(a_n, a_{n-1} \dots a_0 )_r = \sum _{i=0}^n a_i r^i = a_0 + r \sum_{i=1} ^n a_i r^{i-1} $$

Hence, $a_0$ is simply **the remainder of $A_r$ divided by $r$**.

To find $a_1$, we simply consider
$$A'_r = (a_n, a_{n-1} \dots a_1)_r = \left \lfloor \frac{A_r}{r}\right \rfloor$$
And we compute $a_1 = a'_0$.

By repeating this process, we can find all the $a_i$ to left of the radix point.

For the right of the radix point, consider the number if we remove the parts before the radix point.
$$(a_{-1}, a_{-2} \dots a_{-m} )_r = \sum _{i=-m}^{-1} a_i r^i = a_{-1} r^{-1} + r^{-2} \sum_{i=-m} ^{-2} a_i r^{i+2} $$

Now suppose that we multiply the number by $r$.
$$r \sum _{i=-m}^{-1} a_i r^i = r \left(a_{-1} r^{-1} + r^{-2} \sum_{i=-m} ^{-2} a_i r^{i+2}\right) = a_{-1} + r^{-1} \sum_{i=-m} ^{-2} a_i r^{i+2}  $$

Notice that the second term is always $< 0$, hence $a_{-1}$ is the integer part of the resultant value.

Similarly, we can find all of $a_i$ to the right of the radix point by repeatedly multiplying by 2 and extracting the integer part.

In [80]:
def from_dec(n: float, r: int, max_len = 10):
    SYMBOLS = '0123456789ABCDEF'
    
    left = int(n)
    right = n - left
    
    l_digits = []
    while left:
        l_digits.append(SYMBOLS[left % r])
        left //= r
        
    r_digits = []
    for _ in range(max_len):
        if right == 0:
            break
        right *= r
        r_digits.append(SYMBOLS[int(right)])
        right -= int(right)
    
    return ''.join((*reversed(l_digits), '.', *r_digits))

In [81]:
print(from_dec(13.75, 2))
print(from_dec(13.75, 16))
print(from_dec(13.75, 8))

1101.11
D.C
15.6


## Between binary, hex and octal

### Hex $\leftrightarrow$ Binary

Each hex digit $\leftrightarrow$ 4 binary digits

### Octal $\leftrightarrow$ Binary

Each octal digit $\leftrightarrow$ 3 binary digits

### Hex $\leftrightarrow$ Octal

Use binary as an intermediate step

---

For all the steps above, remember to group the digits starting from the radix point.

## Binary arithmetic

### Addition

Similar to the addition we are used to in the decimal world, we can add digits together, and if they become too large, we "carry" the value to the next position.

The following is the addition table in binary:
```
0 + 0 =  0 
0 + 1 =  1 
1 + 1 = 10 (carry)
```

Thus, for example:

```
  111       Carry
-----------
   10100
+   1110
----------- 
= 100010
```

To verify this, we can check that $$(10100)_2 + (1110)_2 = (20)_{10} + (14)_{10} = (34)_{10} = (100010)_2$$

### Multiplication

To perform basic multiplication in decimal system, we compute each digit of one number multiplied by the second number, shifted by tens according to the digit's position.
Then we sum it up.
For example:

```
-----------
    1254
x    302
----------- 
    2508
       0
+ 376200
-----------
= 378708
```

This is much simpler in the binary case, since the digits is either 0 or 1, we simply just shift if by the relevant amount of spaces if the digit is a 1.

```
------------
     10100
*     1101
------------ 
     10100
   10100
+ 10100
------------
 100000100
```

To verify this, we can check that $$(10100)_2 \times (1101)_2 = (20)_{10} + (13)_{10} = (260)_{10} = (100000100)_2$$

### Subtraction

The following is the subtraction table in binary:
```
0 - 0 = 0 
1 - 0 = 1 
1 - 1 = 0
0 - 1 = 1 (borrow)
```

Thus, for example:

```
   10100
-   1110
----------- 
=    110
```

### Division

For division, we do the same as in the decimal, but by simply shifting and subtracting.

```
       111
     ____________
101 / 100101
   -   101    
    -------------
       10001    
   -    101
    -------------
        0111   
   -     101
    -------------
          10
```

To verify this, we can check that 
$$(100101)_2 = (37)_{10} = (7)_{10} \times (5)_{10} + (2)_{10} = (111)_2 \times (101)_2 + (10)_2$$

## Computer arithmetic

Notice that in binary, we can use addition to implement multiplication, and subtraction to implement division.
Suppose that we can represent negative numbers, then we can perform subtraction by adding a negative number, which allows us to perform all binary arithmetic operations just by using adders.

Hence, our goal is to find a representation of binary numbers which is valid when adding negative numbers.

**Most significant bit**: the left-most bit of a binary number

**Least significant bit**: the right-most bit of a binary number

### Unsigned binary

The binary system that we'd introduce is actually the unsigned binary system.
Notice that we cannot represent negative numbers with it.
Hence, the possible range of a $n$-bit unsigned integer is $[0, 2^n-1]$.

### Signed binary

A straightforward way of representing negative numbers is to reserve the most-significant bit to represent the sign.
$0$ means the number is positive, and $1$ means it is negative.

For example, `1101` would be $-5$.

The possible range of a $n$-bit signed integer is $[-2^{-n}-1, 2^{n-1}-1]$.

Now suppose we try to perform subtraction by adding a negative signed binary number.

1 in binary is `001`, thus -1 in unsigned binary is `101`.

```
   3
-  1
----------- 
=  2
```

```
    011
+   101
----------- 
=(1)000
```

Since we are working with 3-bits, the final carry `1` will be discarded, leaving us with `000`, which is wrong.

Thus, signed binary is unable to fulfill our needs.

### 1's complement

For a number system with radix $r$, the **diminished radix complement** of a $n$-digit number $A$ is defined as:
$$
A^{**} = r^n - A - 1
$$

The 1's complement is the diminished radix complement of a binary number.

Since $r^n - 1$ is simply a $n$-digit number with all 1's, $A^{**}$ is simply $A$ but with all its bits flipped.

For example, $A = (1010)_2, A^{**} = (1111)_2 - (1010)_2 = (0101)_2$

The possible range of a $n$-bit 1's complement is $[-2^{-n}-1, 2^{n-1}-1]$, the same as signed integer.

Now suppose we try to perform subtraction by adding 1's complement representation of the number.

1 in binary is `001`, thus -1 is the 1's complement which is `110`.

```
   011
+  110
----------- 
= 1001
```

For the sum to be correct, we need to add the final carry bit back into the number, thus we get:

```
   011
+  110
----------- 
(1)001
+    1
----------- 
   010
```

Verify that the above is correct in the decimal system.

Similarly, for $3 + (-2)$ in the decimal world:

```
   011
+  101
----------- 
(1)000
+    1
----------- 
   001
```

### 2's complement

For a number system with radix $r$, the **radix complement** of a $n$-digit number $A$ is defined as:
$$
A^{*} = r^n - A = A^{**} + 1
$$

The 2's complement is the radix complement of a binary number.

This is also simply the 1's complement of the number + 1.

The possible range of a $n$-bit 2's complement is $[-2^{-n}, 2^{n-1}-1]$.

Now suppose we try to perform subtraction by adding 2's complement representation of the number.

The 1's complement of -1 which was `110`, hence the 2's complement is `111`.

For the sum to be correct, we ignore the overflow bit.

```
   011
+  111
----------- 
(1)010
```

Similarly, for $3 +(-2)$:

```
   011
+  110
----------- 
(1)001
```

#### Benefits over 1's complement
* No need to process final carry bit
* There is only 1 representation of 0 (`000`), while 1's complement has 2 (`000` and `111`)

## Representation limits

### Overflow

For a $n$-bit number, we know that there is a limit to how large a number it can represent.
It is possible that our addition result in a number that is beyond this limit, this is when an **overflow** occurs.

When an overflow occurs, the result is not valid.

For example, adding `011` and `010` in 2's complement (equivalent to 3 and 2 in decimal) result in `101`, which is -3, giving us a wrong result.

#### Detecting overflows

To detect overflow, we know it happens **when the sign of the two operands are the same, but different from the sign of the result**.

As in our previous example, we are adding two positive number, and yet resulted in an negative number, hence we know that overflow has occurred.